recipes-extended/xen/files/libxl-iso-hotswap.patch

```diff
@@ -0,0 +1,346 @@
+commit d04740f8f1dfc5c2149863aa3f896ac4664b5c88
+Author: Chris Rogers <rogersc@ainfosec.com>
+Date:   Tue Apr 18 14:43:21 2017 -0400
+
+    implement iso hotswap
+
+Index: xen-4.6.4/tools/libxl/libxl.c
+===================================================================
+--- xen-4.6.4.orig/tools/libxl/libxl.c
++++ xen-4.6.4/tools/libxl/libxl.c
+@@ -2940,6 +2940,126 @@ int libxl_device_disk_getinfo(libxl_ctx
+     return rc;
+ }
+ 
++int libxl_cdrom_change(libxl_ctx *ctx, uint32_t domid, char *iso, libxl_device_disk *olddisk, char *vdev,
++                       const libxl_asyncop_how *ao_how)
++{
++    AO_CREATE(ctx, domid, ao_how);
++    int rc = -1;
++    libxl__device device;
++    libxl_device_disk disk_empty;
++    xs_transaction_t t = XBT_NULL;
++    char *devpath = NULL;
++    char *be_path = NULL;
++    char *fe_path = NULL;
++    char *tmp = NULL;
++    char *be_state = NULL;
++    char *fe_state = NULL;
++    uint32_t stubdomid = libxl_get_stubdom_id(ctx, domid);
++    libxl__domain_userdata_lock *lock = NULL;
++
++    memset(&device, 0, sizeof(libxl__device));
++    memset(&disk_empty, 0, sizeof(libxl_device_disk));
++
++    struct xs_permissions roperm[2];
++    roperm[0].id = 0;
++    roperm[0].perms = XS_PERM_NONE;
++    roperm[1].id = stubdomid;
++    roperm[1].perms = XS_PERM_READ;
++
++    /* get our empty disk setup */
++    libxl_device_disk_init(&disk_empty);
++    disk_empty.format = LIBXL_DISK_FORMAT_EMPTY;
++    disk_empty.vdev = libxl__strdup(NOGC, olddisk->vdev);
++    disk_empty.pdev_path = libxl__strdup(NOGC, "");
++    disk_empty.is_cdrom = 1;
++    libxl__device_disk_setdefault(gc, &disk_empty);
++
++    lock = libxl__lock_domain_userdata(gc, domid);
++    if (!lock) {
++        rc = ERROR_LOCK_FAIL;
++        goto out;
++    }
++
++    /* tap iso, if it's already tapped it will return the existing tap */
++    devpath = libxl__blktap_devpath(gc, iso, LIBXL_DISK_FORMAT_EMPTY, "/config/platform-crypto-keys");
++
++    rc = libxl__device_from_disk(gc, stubdomid, olddisk, &device);
++    if (rc){
++        fprintf(stderr, "can't create device from disk\n");
++        goto out;
++    }
++    /* insert empty cdrom */
++    libxl__qmp_insert_cdrom(gc, domid, &disk_empty);
++
++    be_path = libxl__device_backend_path(gc, &device);
++    fe_path = libxl__device_frontend_path(gc, &device);
++
++    /* disconnect iso for now */
++    do {
++        t = xs_transaction_start(ctx->xsh);
++        tmp = libxl__xs_read(gc, t, libxl__sprintf(gc, "%s/tapdisk-params", be_path));
```

Comment:
So these return vals from these calls to libxl__xs_read are not cleaned up. I assume it is because the gc gets them?